### ISSUE TITLE
S3 client unconditionally hits us-west-1 unless you specify otherwise

### ISSUE THREAD

**Issue by mcgodes** [aws/aws-sdk-java#793]
This bites us when we use a VPC endpoint in another region, because us-west-1 is inaccessible. The relevant code appears to be [this](https://github.com/aws/aws-sdk-java/blob/master/aws-java-sdk-s3/src/main/java/com/amazonaws/services/s3/AmazonS3Client.java#L4420-L4431):

```java
    /**
     * Retrieves the region of the bucket by making a HeadBucket request to us-west-1 region.
     *
     * Currently S3 doesn't return region in a HEAD Bucket request if the bucket
     * owner has enabled bucket to accept only SigV4 requests via bucket
     * policies.
     */
    private String getBucketRegionViaHeadRequest(String bucketName) {
        String bucketRegion = null;

        try {
            String endpoint = "https://s3-us-west-1.amazonaws.com";
            Request<HeadBucketRequest> request = createRequest(bucketName, null,
                    new HeadBucketRequest(bucketName), HttpMethodName.HEAD, new URI(endpoint));
```

**Comment by shorea** [comment]
Is it possible to configure the S3 client with an explicit region. We only use this fallback behavior when no region is provided.

**Comment by shorea** [comment]
Feel free to reopen if workaround is suggested is not feasible.

**Comment by chtompki** [comment]
This seems fairly counter intuitive for `us-west-1` to be the default, while historically it has been `us-east-1`. Why not simply throw an exception for no declared region?

**Comment by varunnvs92** [comment]
In all versions of the SDK, we allow clients to be created with no declared region. In 1.11.x, we need region because we do SigV4 signing by default which requires the bucket region.

If we start throwing an exception for no region, we will break a lot of customers code. Also us-west-1 is not default, we do it only once to find the region (via HeadBucket) and cache it. For all subsequent requests to the same bucket, we use the correct region from the cache.

**Comment by chtompki** [comment]
That's fair. Regardless, it still seems odd to go to us-west-1, even for finding the bucket's region, in the undeclared scanario. Is that the same behavior in boto3?

**Comment by varunnvs92** [comment]
I agree its odd that it goes to us-west-1 but we need to hit some endpoint(not s3.amazonaws.com) to find the region, so we chose it. We are working with Amazon S3 team to have better approach instead of HeadBucket approach which has caused few other issues. It may take some time but will let you know if there are any updates.

I am not entirely sure but I think they have a different approach.

**Comment by amoghmargoor** [comment]
is it fair expectation to see this issue go away when we set endpoint explicitly (and not just when region is set explicitly) ? Internally clientRegion gets set due to setting of endpoint but not when endpoint belongs to `us-east-1` (may be because it is default). Should setting of clientRegion always while setting endpoint be done to help us with this issue ?

**Comment by varunnvs92** [comment]
This issue is fixed in the recent versions of 1.11.x. SDK no longer makes cross region calls.

>>  Should clientRegion be always set while setting endpoint ?
No need to set both. You either need to set the region or set the endpoint along with signer override region. I suggest you to use fluent client builders which make the usage experience clear.
